### ISSUE TITLE
add weekFormatter just like dayOfWeekFormatter

### ISSUE THREAD

**Issue by amakaa** [mui/mui-x#9785]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

requesting a prop like dayOfWeekFormatter for the week numbers

### Examples 🌈

https://github.com/search?q=repo%3AHacker0x01%2Freact-datepicker%20formatweeknumber&type=code

### Motivation 🔦

I am trying to change the week numbers but not able to. using a different way of calculating week numbers (for organization) that differs from the current standard formats

**Comment by noraleonte** [comment]
Thank you for you suggestion! This is definitely something worth exploring within the team - which we will do. 🚀

**Comment by flaviendelangle** [comment]
Are you trying to customize the week number displayed in [this doc section](https://mui.com/x/react-date-pickers/date-calendar/#display-week-number)

If so, you can use the `calendarWeekNumberText` locale text key
More information on how to set a custom locale text key [on this doc section](https://mui.com/x/react-date-pickers/localization/#set-translations-locally)

**Comment by amakaa** [comment]
> Are you trying to customize the week number displayed in [this doc section](https://mui.com/x/react-date-pickers/date-calendar/#display-week-number)
> 
> If so, you can use the `calendarWeekNumberText` locale text key More information on how to set a custom locale text key [on this doc section](https://mui.com/x/react-date-pickers/localization/#set-translations-locally)

it is close to what i'm looking for but because the only arg is weekNumber, it doesn't work for changing the week number to another value by using the dates to recalculate. so for this i would need access to the date range or array of dates for that particular week or calculate based on the day (from the example linked) for example

**Comment by flaviendelangle** [comment]
If you need the date, can't you build it from the week number ?

**Comment by amakaa** [comment]
> If you need the date, can't you build it from the week number ?

i didn't even know that was possible. thanks for the suggestion, i will try that

**Comment by flaviendelangle** [comment]
If you know the year, you should be able to do it
If not, we could add other params to the callback :+1:

**Comment by amakaa** [comment]
> If you know the year, you should be able to do it If not, we could add other params to the callback 👍

yes, i believe the additional params would still be helpful. i am unable to access the year to build the date from the week number

**Comment by flaviendelangle** [comment]
@alexfauquette would the following API make sense to you?

```diff
diff --git a/packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx b/packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx
index a256dff17..1a8f7de1f 100644
--- a/packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx
+++ b/packages/x-date-pickers/src/DateCalendar/DayCalendar.tsx
@@ -603,7 +603,9 @@ export function DayCalendar<TDate>(inProps: DayCalendarProps<TDate>) {
                       utils.getWeekNumber(week[0]),
                     )}
                   >
-                    {localeText.calendarWeekNumberText(utils.getWeekNumber(week[0]))}
+                    {localeText.calendarWeekNumberText(utils.getWeekNumber(week[0]), {
+                      firstDay: week[0],
+                    })}
                   </PickersCalendarWeekNumber>
                 )}
                 {week.map((day, dayIndex) => (
diff --git a/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts b/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
index a2acb4fa2..401114b07 100644
--- a/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
+++ b/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
@@ -33,7 +33,7 @@ export interface PickersComponentAgnosticLocaleText<TDate> {
   calendarWeekNumberHeaderLabel: string;
   calendarWeekNumberHeaderText: string;
   calendarWeekNumberAriaLabelText: (weekNumber: number) => string;
-  calendarWeekNumberText: (weekNumber: number) => string;
+  calendarWeekNumberText: (weekNumber: number, params: { firstDay: TDate }) => string;
 
   // View navigation
   openPreviousView: string;
```

**Comment by alexfauquette** [comment]
Modifying the week number seems to be very strange, but why not. I don't see issues to add an additional parameter

You should also to the modification in `DayCalendar.tsx`

```jsx
aria-label={localeText.calendarWeekNumberAriaLabelText(utils.getWeekNumber(week[0]),)}
```

Another option would be to modify the adapter method `getWeekNumber` that were created especially for this feature

**Comment by flaviendelangle** [comment]
> Another option would be to modify the adapter method getWeekNumber that were created especially for this feature

It's a personal preference, but I think we should not encourage people to customize the adapter.
It's yet another way of doing things which, I think, will add more confusion.

____

@amakaa could you go a little more in depth in your use case ?
What value are you trying to output ? 
Maybe we are missing another solution

**Comment by amakaa** [comment]
Sure. From the example I linked from `react-datepicker`, the date is passed into the week formatter and this helps with the use case of recalculating the week number. It is done in this case to ensure that the last week number of the year stops at the week that contains the last day of the year. The week number then restarts from 1 again after this. This first screenshot is default week numbers
<img width="309" alt="Screenshot 2023-08-01 at 14 46 21" src="https://github.com/mui/mui-x/assets/2635534/a8f3b6bb-c5ba-4653-bc93-9f86324ca7e1">
This second screenshot is with the week numbers recalculated
<img width="310" alt="Screenshot 2023-08-01 at 14 34 07" src="https://github.com/mui/mui-x/assets/2635534/82fcbbea-72d0-4bad-871f-c3d1fd027bc7">

**Comment by alexfauquette** [comment]
But that's not what we do 🤔

Here for example the week number starts back at 1 without having to do any modification.
https://codesandbox.io/s/sad-julien-t7kddl?file=/Demo.tsx

![image](https://github.com/mui/mui-x/assets/45398769/d39c1b96-fc7c-4847-9c8b-902ca8a0a20c)


Defining when is the first week is a bit more challenging since there is multiple definitions of what is the first week of the year (see [wikipedia](https://en.wikipedia.org/wiki/Week#Other_week_numbering_systems) for more details) So it's more a configuration to do on the date library used.

**Comment by amakaa** [comment]
Right, so the complicated part is this - the week system that would work this way would be ISO 8601 (which starts on a Monday). However, in this use case, the desire is for the week to start on a Sunday so it complicates things a bit. But since this is an edge case then understandable if it's not worth including the date param. OK with closing the issue.

**Comment by alexfauquette** [comment]
> However, in this use case, the desire is for the week to start on a Sunday

Interesting, date libraries are for sure allowing you to choose which day is the first day of the week. But I don't know how it affects the week numbering. Would not be surprised if it was not consistent across libraries 🙈

Maybe you could specify what you were expecting:
- week starting day: Sunday
- week number 1 definition: ... ?

And the reason why: is it related to your localization, or specific to some domain (finance, military, construction, ...)

Then we could try to see what can be done from libraries and see if we can lose the issue. For now, I put the "waiting upvote" to see if others have similar issue

**Comment by amakaa** [comment]
thanks for following up.

expecting:

- uses IS8601 format
- but week to start on Sunday instead

it's specific to the domain of retail

**Comment by alexfauquette** [comment]
I updated the demo to test multiple stuff

https://codesandbox.io/s/sad-julien-t7kddl?file=/Demo.tsx

Basically with dayjs you can do it without issue thanks to [locale customization](https://day.js.org/docs/en/customization/customization).

```jsx
dayjs.extend(updateLocale);

dayjs.updateLocale("en", {
  weekStart: 1, // OPTIONAL, set the start of a week. If the value is 0, Monday will be the start of week instead of Sunday。
  yearStart: 4, // OPTIONAL, the week that contains Jan 4th is the first week of the year.
});
```

For some other libraries it is not possible. For example Luxon stays with ISO week definition

**Comment by amakaa** [comment]
it also works with `date-fns`

```
import { setDefaultOptions } from 'date-fns';
setDefaultOptions({ weekStartsOn: 0, firstWeekContainsDate: 4 });
```

**Comment by alexfauquette** [comment]
Nice, then I propose to add a message in the docs saying that week numbers come from the date library configuration, and give those example about how to customize it before closing this issue

**Comment by roelvdwater** [comment]
It is also possible with `moment`:

```
import moment from "moment";

moment.updateLocale("en", {
  week: {
    dow: 1,
    doy: 4
  }
});
```

More information can be found in [their documentation](https://momentjscom.readthedocs.io/en/latest/moment/07-customization/16-dow-doy/).

**Comment by LukasTy** [comment]
As the adapters do support changing (overriding the localized one) the start of the week, we've decided to pivot this issue into a documentation improvement.
The idea is to document how the user can override the start of the week with different adapters.
